core/server/worker/src/main/java/alluxio/worker/grpc/BlockWriteHandler.java

```diff
@@ -103,6 +110,7 @@ protected void cleanupRequest(BlockWriteRequestContext context) throws Exception
       context.getBlockWriter().close();
     }
     mWorker.cleanupSession(context.getRequest().getSessionId());
+    RPC_WRITE_COUNT.dec();
```

Comment:
I wasn't able to tell from the code, but I think I can modify it and test whether or not that occurs. Will get back to this comment once I've done that test.

Reply:
I added log statements to each of those blocks, and here was the result of running `./bin/alluxio fs -Dalluxio.user.short.circuit.enabled=false copyFromLocal ./LICENSE /LICENSE`:
```
2021-05-25 18:47:14,240 INFO  BlockWriteHandler - [czhu] WRITE createRequestContext
2021-05-25 18:47:14,244 INFO  BlockWriteHandler - [czhu] WRITE completeRequest
2021-05-25 19:40:33,333 INFO  JvmPauseMonitor - JVM paused 1574ms
No GCs detected 
max memory = 3641M total memory = 148M free memory = 60M
2021-05-25 19:44:13,730 INFO  JvmPauseMonitor - JVM paused 1547ms
No GCs detected 
max memory = 3641M total memory = 148M free memory = 58M
```

Reply:
This could be an issue on cancel path, regular completion will be clean afaict.

Reply:
https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/StreamObserver.html
will provide insights into how these handlers interact with clients...

Reply:
From the `StreamObserver` interface, it says either of `onCompleted()` or `onError()` must be the last method called.

`BlockWriteHandler` extends `AbstractWriteHandler` which has [the following calls](https://github.com/Alluxio/alluxio/blob/master/core/server/worker/src/main/java/alluxio/worker/grpc/AbstractWriteHandler.java#L401) for `mResponseObserver`:
```
  /**
   * Writes a response to signify the success of the write request.
   */
  private void replySuccess() {
    mContext.setDoneUnsafe(true);
    mResponseObserver.onCompleted();
  }

  /**
   * Writes a response to signify the successful cancellation of the write request.
   */
  private void replyCancel() {
    mContext.setDoneUnsafe(true);
    mResponseObserver.onCompleted();
  }

  /**
   * Writes an error response.
   */
  private void replyError() {
    replyError(Preconditions.checkNotNull(mContext.getError()));
  }

  /**
   * Writes an error response.
   */
  private void replyError(Error error) {
    if (error.isNotifyClient()) {
      mResponseObserver.onError(error.getCause().toGrpcStatusException());
    }
  }
```

So I agree `BlockWriteHandler#cancelRequest()` is probably not supposed to decrement the counter.

Reply:
It seems, this will be called when complete/cancel has trouble. Depending on the place of error, this could cause decrementing the counter twice. (If `replySuccess()` or `replyCancel()` fails)

So ideally each write-handler should remember its increment/decrement status in a boolean flag (if can't reuse an existing one) and don't decrement here if it has already been decremented.

Reply:
Flag has been added, additionally from `AbstractWriteHandler`:
```
  /**
   * Completes this write. This is called when the write completes.
   *  @param context context of the request to complete
   */
  protected abstract void completeRequest(T context) throws Exception;

  /**
   * Cancels this write. This is called when the client issues a cancel request.
   *
   * @param context context of the request to complete
   */
  protected abstract void cancelRequest(T context) throws Exception;

  /**
   * Cleans up this write. This is called when the write request is aborted due to any exception
   * or session timeout.
   *
   * @param context context of the request to complete
   */
  protected abstract void cleanupRequest(T context) throws Exception;
```

I can't tell whether or not `cancelRequest()` results in `completeRequest()` getting called too from the description. The implementation with the `StreamObserver` with `replyCancel()` seems to imply Cancel -> Completed, but that may just be for the `StreamObserver`.